Add standardized messages and tools attributes in span schema

daniellok-db opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/daniellok-db/mlflow/pull/14087?quickstart=1)

#### Install mlflow from this PR

```
# Use `%sh` to run this command on Databricks
OPTIONS=$(if pip freeze | grep -q 'mlflow @ git+https://github.com/mlflow/mlflow.git'; then echo '--force-reinstall --no-deps'; fi)
pip install $OPTIONS git+https://github.com/mlflow/mlflow.git@refs/pull/14087/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 14087
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

This PR adds the `mlflow.chat.messages` and `mlflow.chat.tools` attributes to the span schema.

We reuse the MLflow AI Gateway definitions, and extend the function definitions to support the full range of OpenAI definitions.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

Existing unit tests pass, and added some new tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [x] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Added a schema for standardized chat model messages and tool definitions in `LLM` / `CHAT_MODEL` spans.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

daniellok-db committed:
initial commit

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
property

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
test

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db commented on mlflow/utils/__init__.py:
```diff
@@ -6,11 +6,16 @@
 
 PYTHON_VERSION = f"{version_info.major}.{version_info.minor}.{version_info.micro}"
 
+IS_PYDANTIC_V1 = Version(pydantic.__version__).major < 2
+
```
i'm not sure this is the best place for it, but i was running into circular import issues elsewhere

daniellok-db committed:
test

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db commented on mlflow/utils/__init__.py:
```diff
@@ -6,12 +6,15 @@
 PYTHON_VERSION = f"{version_info.major}.{version_info.minor}.{version_info.micro}"
 
 
+IS_PYDANTIC_V2 = Version(pydantic.__version__).major >= 2
```
not sure if this file is the best place for it, but i noticed we define some version of this constant in many places and i wanted to unify

harupy replied:
```suggestion
IS_PYDANTIC_V2_OR_NEWER = Version(pydantic.__version__).major >= 2
```

Let's be more precise here :)

daniellok-db commented on tests/gateway/providers/test_mosaicml.py:
```diff
@@ -478,26 +479,6 @@ def test_valid_parsing(messages, expected_output):
     )
 
 
-@pytest.mark.parametrize(
-    "messages",
-    [
-        [RequestMessage(role="invalid_role", content="Test")],
-        [RequestMessage(role="another_invalid_role", content="Test")],
-        [
-            RequestMessage(role="system", content="Test"),
-            RequestMessage(role="user", content="Test"),
-            RequestMessage(role="invalid_role", content="Test"),
-        ],
-    ],
-)
-def test_invalid_role_submitted_raises(messages):
-    route_config = RouteConfig(**chat_config())
-    with pytest.raises(
-        MlflowException, match=".*Must be one of 'system', 'user', or 'assistant'.*"
-    ):
-        MosaicMLProvider(route_config)._parse_chat_messages_to_prompt(messages)
-
-
```
it is not possible to construct a `RequestMessage` with an invalid role now

daniellok-db committed:
fix

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
test

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db commented on mlflow/metrics/genai/model_utils.py:
```diff
@@ -145,7 +145,7 @@ def _call_llm_provider_api(
     payload = {
         k: v
-        for k, v in chat_request.model_dump().items()
+        for k, v in chat_request.model_dump(exclude_none=True).items()
```
if exclude_none is not set, some keys may be present which don't belong (e.g. tool_calls in user messages)

daniellok-db commented on mlflow/gateway/schemas/chat.py:
```diff
@@ -1,17 +1,119 @@
+
     role: str
-    content: str
+    content: Optional[ContentType] = None
+    tool_calls: Optional[list[ToolCall]] = Field(None, min_items=1)
+    tool_call_id: Optional[str] = None
+    refusal: Optional[str] = None
```
i wish there was a way to simply define a type union like

```
RequestMessage = Union[UserMessage, SystemMessage, AssistantMessage, ToolMessage]
```

but it seems like every way of doing this has some problems. If we simply define it as a type (see Cat and Pet [in this example](https://docs.pydantic.dev/latest/concepts/unions/#nested-discriminated-unions)), then we can't do `type.model_validate()`, because it is not a class.

if we [extend RootModel](https://docs.pydantic.dev/latest/concepts/unions/#nested-discriminated-unions), then we can't access fields like `message.role` (and this RootModel class does not exist in Pydantic 1)

B-Step62 replied:
I agree with this approach and indeed don't think we wanna separate subclasses for types. The goal is to be OpenAI compatible and they define [the schema](https://github.com/openai/openai-python/blob/e94d98e9bf97a5d2d02d79d58f2abdbab26ff2bd/src/openai/types/chat/chat_completion_message.py#L26) as a unified chat message class. The design of having subclasses per type seems to be Langchain-unique.

B-Step62 commented on docs/source/llms/tracing/tracing-schema.rst:
```diff
@@ -331,3 +331,45 @@ documents from a vector store). The ``RETRIEVER`` span type has the following sc
 \* For example, both ``[Document(page_content="Hello world", metadata={"doc_uri": "https://example.com"})]`` and
 ``[{"page_content": "Hello world", "metadata": {"doc_uri": "https://example.com"}}]`` are valid outputs for a ``RETRIEVER`` span.
+
+Chat Completion Spans
```
Can we add a quick example of adding those attributes to a manually created span? I think it makes it more clearer that we use custom attributes here not the actual input/output.

B-Step62 commented on mlflow/entities/span.py:
```diff
@@ -144,6 +155,16 @@ def outputs(self) -> Any:
         """The output values of the span."""
         return self.get_attribute(SpanAttributeKey.OUTPUTS)
 
+    @property
+    def messages(self) -> str:
+        """The chat messages of the span, if any"""
+        return self.get_attribute(SpanAttributeKey.CHAT_MESSAGES)
```
Hmm I'm reluctant to add these properties to the Span object. There is no future limit for what type of spans we will support, and adding specific arguments for each type will make this class dirty. I think asking users to call `get_attribute(spanAttributeKey.CHAT_MESSAGES` is fair as everything used there is public?

daniellok-db replied:
yeah that's fair, it may get quite crowded

B-Step62 commented on mlflow/entities/span.py:
```diff
@@ -308,6 +329,41 @@ def set_outputs(self, outputs: Any):
         """Set the output values to the span."""
         self.set_attribute(SpanAttributeKey.OUTPUTS, outputs)
 
+    def set_chat_messages(
```
With the same reason above, can we move this out to separate utility? We don't want to have a type-specific logic within the general Span class.

daniellok-db replied:
makes sense, what do you think about some fluent api like `mlflow.set_span_chat_messages(span, messages)`?

B-Step62 replied:
Yea the API sounds good, I'd put it in `mlflow.tracing.xyz` not to overload `mlflow` namespace tho

B-Step62 replied:
Another option is to put it into the model side:
```
mlflow.set_attributes(messages.to_span_attribute())
```
Probably the message model is less crowded than span model (cuz it is purpose built)

daniellok-db replied:
i guess there's no harm, anyway probably most users will be using autologging so it shouldn't add too much friction

B-Step62 commented on mlflow/gateway/schemas/chat.py:
```diff
@@ -1,17 +1,119 @@
+    text: str
+
+
+class ImageUrl(RequestModel):
```
Is there any reason we use `RequestModel` rather than the `BaseModel`? 

The docstring of the `BaseModel` says
```
    # Allow extra fields for pydantic request models, e.g. to support
    # vendor-specific embeddings parameters
```

I don't think we need to support extra fields here? If possible, less number of abstraction is better.

daniellok-db replied:
ah perfect, let me switch

B-Step62 commented on mlflow/tracing/constant.py:
```diff
@@ -20,6 +20,8 @@ class SpanAttributeKey:
     SPAN_TYPE = "mlflow.spanType"
     FUNCTION_NAME = "mlflow.spanFunctionName"
     START_TIME_NS = "mlflow.spanStartTimeNs"
+    CHAT_MESSAGES = "mlflow.chat.messages"
+    CHAT_TOOLS = "mlflow.chat.tools"
```
nit: Can we put a brief comment about what these attributes are used for?

daniellok-db committed:
improve docs and address comments

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix link

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

harupy commented on mlflow/tracing/fluent.py:
```diff
@@ -792,3 +794,109 @@ def _merge_trace(
+    """
+    try:
+        for tool in tools:
+            if IS_PYDANTIC_V2:
+                FunctionTool.model_validate(tool)
+            else:
+                FunctionTool.parse_obj(tool)
```
Can we extend `BaseModel` to add a helper method for validation? 

```python
import pydantic
from pydantic import BaseModel as BaseModel_
from packaging.version import Version
from typing import Any

IS_PYDANTIC_V2 = Version(pydantic.__version__).major >= 2


class BaseModel(BaseModel_):
    @classmethod
    def validate(cls, obj: Any):
        if IS_PYDANTIC_V2:
            cls.model_validate(obj)
        else:
            cls.parse_obj(obj)


BaseModel.model_validate(...)
```

harupy commented on mlflow/gateway/schemas/chat.py:
Is this really the right place to define these classes? Can we move them in `mlflow/types`?

harupy commented on mlflow/tracing/fluent.py:
```diff
@@ -792,3 +794,109 @@ def _merge_trace(
+                RequestMessage.model_validate(message)
+            else:
+                RequestMessage.parse_obj(message)
+    except pydantic.ValidationError as e:
+        raise MlflowException.invalid_parameter_value(
+            "Chat messages must be a list of valid messages."
+        ) from e
+    span.set_attribute(SpanAttributeKey.CHAT_MESSAGES, messages)
```
Is there any method in pydantic we can extend to implement this error conversion?

daniellok-db replied:
actually i think we can just raise the base pydantic validation error, i don't think there's a benefit to doing the conversion

harupy replied:
When this code is running in a server, the error type (= status code) is important.

daniellok-db replied:
i think these functions should only be called in clients (since they operate on livespans), but added back the conversion as there should be no harm. also added a couple tests for validation

harupy commented on mlflow/gateway/schemas/chat.py:
```diff
@@ -1,17 +1,119 @@
-from typing import Literal, Optional
+from typing import Literal, Optional, Union
```
```suggestion
from __future__ import annotations

from typing import Literal, Optional, Union
```

We can use `from __future__ import annotations` to simplify type annotation using [PEP604](https://peps.python.org/pep-0604/)

harupy replied:
@daniellok-db Filed https://github.com/mlflow/mlflow/pull/14110. Once this PR is merged, we can just run `ruff check --unsafe-fixes --fix .` to automatically transform the code.

daniellok-db replied:
just tried this and it looks like it doesn't play nicely with pydantic. basemodel doesn't seem to allow the use of `|` operator

daniellok-db committed:
comments

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
oops

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
add back exceptions

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

harupy reviewed (approved):
LGTM!

harupy commented on mlflow/types/chat.py:
```diff
@@ -0,0 +1,130 @@
+
+from pydantic import BaseModel as _BaseModel
+from pydantic import Field
+from typing_extensions import Annotated
```
MLflow doesn't depend on `typing_extensions`. We should add it as a dependency if we use it.

harupy replied:
In which python version was `Annotated` introduced?

harupy commented on mlflow/types/chat.py:
```diff
@@ -0,0 +1,130 @@
+            Union[TextContentPart, ImageContentPart, AudioContentPart], Field(discriminator="type")
+        ]
+    ],
+    Field(min_items=1),
```
Do we need `min_items=1`? The following query is useless but works fine.


```
$curl https://api.openai.com/v1/chat/completions \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer $OPENAI_API_KEY" \
  -d '{
    "model": "gpt-4o",
    "messages": [
      {
        "role": "user",
        "content": []
      }
    ],
    "max_tokens": 300
  }'
```

daniellok-db replied:
hmm i guess we don't, i'll remove it. weird that it's an error for tools though:

```
curl https://api.openai.com/v1/chat/completions \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer $OPENAI_API_KEY" \
  -d '{
    "model": "gpt-4o",
    "messages": [
      {
        "role": "user",
        "content": []
      }
    ],
    "tools": [],
    "max_tokens": 300
  }'
{
  "error": {
    "message": "Invalid 'tools': empty array. Expected an array with minimum length 1, but got an empty array instead.",
    "type": "invalid_request_error",
    "param": "tools",
    "code": "empty_array"
  }
}
```

harupy commented on mlflow/types/chat.py:
```diff
@@ -0,0 +1,130 @@
+
+class BaseModel(_BaseModel):
+    @classmethod
+    def validate(cls, obj: any):
```
```suggestion
    def validate(cls, obj: Any):
```

Interesting. This should be detected by our custom linter.

harupy commented on docs/source/llms/tracing/tracing-schema.rst:
```diff
@@ -331,3 +331,117 @@ documents from a vector store). The ``RETRIEVER`` span type has the following sc
+        and will also be used in MLflow evaluation in the future. 
+        
+        The type should be ``List[`` :py:class:`RequestMessage <mlflow.types.chat.RequestMessage>` ``]``,
+        following the standardized chat message format from `MLflow AI Gateway <https://mlflow.org/docs/latest/llms/deployments/index.html>`_
```
do we need to mention `MLflow AI Gateway`?

daniellok-db replied:
ah my bad i forgot to edit this when i moved it to `mlflow/types`, we don't have to mention it, but I wanted to indicate that we're following _some_ kind of standard, without having to mention OpenAI directly in case eventually we diverge

harupy commented on docs/source/llms/tracing/tracing-schema.rst:
```diff
@@ -331,3 +331,117 @@ documents from a vector store). The ``RETRIEVER`` span type has the following sc
+  import mlflow
+  from mlflow.entities.span import SpanType
+  from mlflow.tracing.constant import SpanAttributeKey
+  from mlflow.tracing.utils import set_span_chat_messages, set_span_chat_tools
```
```suggestion
  from mlflow.tracing import set_span_chat_messages, set_span_chat_tools
```

Can we add `set_span_chat_messages` and `set_span_chat_tools` in `mlflow/tracing/__init__.py`?

harupy commented on docs/source/llms/tracing/tracing-schema.rst:
```diff
@@ -331,3 +331,117 @@ documents from a vector store). The ``RETRIEVER`` span type has the following sc
+        conversation with the chat model. It enables rich conversation rendering in the UI,
+        and will also be used in MLflow evaluation in the future. 
+        
+        The type should be ``List[`` :py:class:`RequestMessage <mlflow.types.chat.RequestMessage>` ``]``,
```
```suggestion
        The type must be ``List[`` :py:class:`RequestMessage <mlflow.types.chat.RequestMessage>` ``]``,
```

`should` sounds like it's a recommendation, not a requirement.

harupy commented on mlflow/tracing/utils/__init__.py:
```diff
@@ -232,3 +234,115 @@ def exclude_immutable_tags(tags: dict[str, str]) -> dict[str, str]:
+
+    try:
+        for message in messages:
+            RequestMessage.validate(message)
+    except ValidationError as e:
+        raise MlflowException.invalid_parameter_value(
+            "Received invalid chat messages. Please see the pydantic "
+            "error message above for more details"
+        ) from e
```
Can we update `validate` to perform error conversion?

daniellok-db committed:
comments

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

harupy commented on docs/source/llms/tracing/tracing-schema.rst:
```diff
@@ -331,3 +331,116 @@ documents from a vector store). The ``RETRIEVER`` span type has the following sc
+
+      combined_messages = messages + [response]
+
+      span = mlflow.get_current_active_span()
+      set_span_chat_messages(span, combined_messages)
+      set_span_chat_tools(span, tools)
```
```suggestion
      span = mlflow.get_current_active_span()
      span.set_chat_messages(span, combined_messages)
      span.set_chat_tools(span, tools)
```

Is there a reason we don't these functions as an instance method of `Span`?

daniellok-db replied:
see https://github.com/mlflow/mlflow/pull/14087#discussion_r1887867532 for context, we don't want to crowd the root Span object with schema-specific functions

daniellok-db committed:
comments

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
update tests

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
revert automodule

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
comment and fix chat tool definition

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix tests and docs

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

daniellok-db committed:
fix

Signed-off-by: Daniel Lok <daniel.lok@databricks.com>

dsmilkov commented on mlflow/types/chat.py:
curious if we have plans to consolidate namespaces since as a developer if I have an agent that does retrieval, tool calls, and implements `ChatModel/Agent` I have to import types from 3 different places:

- `from mlflow.entities import Document`
     - so I can trace retrieved docs
- `from mlflow.types.chat import RequestMessage, ChatTools`
    - to trace tool calls and messages
- `from mlflow.types.llm import ChatCompletionRequest, ChatCompletionResponse`
    - to implement the input/output interface

daniellok-db commented:
oh man supporting pydantic 1 and 2 is annoying—changes since last review:

1. Got rid of tool union (there is a bug in pydantic 1 that will never be fixed)
2. Changed name of `BaseModel.validate` to `validate_compat`, because fastAPI calls this function and we override it, causing a test failure

B-Step62 reviewed (approved):
LGTM!!
